be/src/exec/pipeline/runtime_filter_types.h

```diff
@@ -0,0 +1,336 @@
+    // local runtime in-filter
+    RuntimeInFilters _in_filters;
+    // global/local runtime bloom-filter(including max-min filter)
+    RuntimeBloomFilters _bloom_filters;
+};
+
+class RuntimeFilterHolder {
+public:
+    void add_collector(RuntimeFilterCollectorPtr&& collector) {
+        _collector_ownership = std::move(collector);
+        _collector.store(_collector_ownership.get(), std::memory_order_release);
+    }
+    RuntimeFilterCollector* get_collector() { return _collector.load(std::memory_order_acquire); }
+    bool is_ready() { return get_collector() != nullptr; }
+
+private:
+    RuntimeFilterCollectorPtr _collector_ownership;
+    std::atomic<RuntimeFilterCollector*> _collector;
+};
+
+// RuntimeFilterHub is a mediator that used to gather all runtime filters generated by HashJoinBuildOperator instances.
+// It has a RuntimeFilterHolder for each HashJoinBuilder instance, when total runtime filter is generated, then it is
+// added into RuntimeFilterHub; the operators consuming runtime filters inspect RuntimeFilterHub and find out its bounded
+// runtime filters. RuntimeFilterHub is reserved beforehand, and there is no need to use mutex to guard concurrent access.
+class RuntimeFilterHub {
+public:
+    void add_holder(TPlanNodeId id) { _holders.emplace(std::make_pair(id, std::make_unique<RuntimeFilterHolder>())); }
+    void add_collector(TPlanNodeId id, RuntimeFilterCollectorPtr&& collector) {
+        get_holder(id)->add_collector(std::move(collector));
+    }
+
+    RuntimeBloomFilters& get_bloom_filters(TPlanNodeId id) {
+        return get_holder(id)->get_collector()->get_bloom_filters();
+    }
+
+    void close_all_in_filters(RuntimeState* state) {
+        for (auto& [_, holder] : _holders) {
+            if (auto* collector = holder->get_collector()) {
+                for (auto& in_filter : collector->get_in_filters()) {
+                    in_filter->close(state);
+                }
+            }
+        }
+    }
+
+    std::vector<RuntimeFilterHolder*> gather_holders(std::set<TPlanNodeId> ids) {
+        std::vector<RuntimeFilterHolder*> holders;
+        holders.reserve(ids.size());
+        for (auto id : ids) {
+            holders.push_back(get_holder(id).get());
+        }
+        return holders;
+    }
+
+private:
+    RuntimeFilterHolderPtr& get_holder(TPlanNodeId id) {
+        auto it = _holders.find(id);
+        DCHECK(it != _holders.end());
+        return it->second;
+    }
+    // Each HashJoinBuildOperatorFactory has a corresponding Holder indexed by its TPlanNodeId.
+    std::unordered_map<TPlanNodeId, RuntimeFilterHolderPtr> _holders;
+};
+
+// RuntimeFilterProbeCollector::do_evaluate function apply runtime bloom filter to Operators to filter chunk.
+// this function is non-reentrant, variables inside RuntimeFilterProbeCollector that hinder reentrancy is moved
+// into RuntimeBloomFilterEvalContext and make do_evaluate function can be called concurrently.
+struct RuntimeBloomFilterEvalContext {
```

Comment:
Do we really need this struct?

Reply:
it is required, RuntimeFilterProbeCollector::do_evaluate is not reentrant, internal variables are accessed.  in pipeline engine, when dop>1, more than one Operator share the RuntimeFilterProbeCollector instance kept in its OperatorFactory and will invoke RuntimeFilterProbeCollector::do_evaluate concurrently, so variables preventing reentrancy should passed in from outside.

Reply:
it is required, RuntimeFilterProbeCollector::do_evaluate is not reentrant, internal variables are accessed.  in pipeline engine, when dop>1, more than one Operator share the RuntimeFilterProbeCollector instance kept in its OperatorFactory and will invoke RuntimeFilterProbeCollector::do_evaluate concurrently, so variables preventing reentrancy should passed in from outside.